include/libhal-soft/i2c_bit_bang.hpp

```diff
@@ -0,0 +1,210 @@
+// Copyright 2024 Khalil Estell
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <chrono>
+#include <libhal/i2c.hpp>
+#include <libhal/output_pin.hpp>
+#include <libhal/units.hpp>
+
+namespace hal {
+/**
+ * @brief A bit bang implementation for i2c. This implementation of i2c only
+ * needs 2 gpio's to work correctly. In its current implementation we only
+ * support single controller multiple peripheral. But have intentions of
+ * supporting multiple controller soon. Additionally, it's important to note
+ * that bit bang is a best try implementation of i2c. As in it will do it's best
+ * to reach the requested frequency but will typically be slower then requested
+ * with brief testing it seems that the error is about 2us per period for the
+ * lpc4078 , so you can try to account for this error with the frequency you set
+ * for example, achieving the correct period for 100kHz would result in setting
+ * slightly more then a 125kHz frequency. Additionally, the fastest frequency
+ * this bit bang can handle is around 200kHz.
+ */
+class i2c_bit_bang : public i2c
+{
+public:
+  /// holds all of the information for an i2c bus
+  struct bus_info
+  {
+    float duty_cycle = 0.5f;
+  };
+
+  struct i2c_pins
+  {
+    output_pin* sda;
+    output_pin* scl;
+  };
+
+  /**
+   * @brief Construct a new i2c bit bang object
+   *
+   * @param p_pins this holds both scl and sda to be used inside of the driver
+   * @param p_steady_clock a steady clock that should have a higher frequency
+   * then the configured frequency for the bit bang
+   * @param p_bus the bus information which will be used which configuring the
+   * bit bang interface
+   */
+
+  i2c_bit_bang(i2c_pins p_pins, steady_clock& p_steady_clock, bus_info p_bus);
```

Comment:
```suggestion
  i2c_bit_bang(i2c_pins p_pins, steady_clock& p_steady_clock, const bus_info& p_bus);
```

Take bus info as a const reference. This is ABI backwards compatible if we decide to expand this struct.